packages/zoe/src/contractFacet/types-ambient.d.ts

```diff
@@ -217,9 +217,15 @@ type OfferHandler<OR extends unknown = unknown, OA = never> =
   | {
       handle: HandleOffer<OR, OA>;
     };
-type ContractMeta = {
-  customTermsShape?: import('@endo/pass-style').CopyRecord<any> | undefined;
-  privateArgsShape?: import('@endo/pass-style').CopyRecord<any> | undefined;
+type ContractMeta<
+  SF extends // import inline to maintain ambient mode
+    import('../zoeService/utils').ContractStartFunction = import('../zoeService/utils').ContractStartFunction,
+> = {
+  customTermsShape?: Record<
+    Parameters<SF>[0] extends ZCF<infer CT> ? keyof CT : never,
+    Pattern
+  >;
+  privateArgsShape?: Record<keyof Parameters<SF>[1], Pattern>;
   upgradability?: 'none' | 'canBeUpgraded' | 'canUpgrade' | undefined;
```

Comment:
aside: What's the difference between `canBeUpgraded` and `canUpgrade`?

While working on [contract upgrade docs](https://docs.agoric.com/guides/zoe/contract-upgrade.html), I couldn't find any definitions of these... not even any distinctions made in code.

Reply:
good question. I had to read through the code myself to refresh. I added docs on the type